File: BigSortingSolution.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <string>
#include <math.h>
#include <regex>

bool compareBySizeAndValue(const std::string& a, const std::string& b);
void createArray(int n);
void display(std::vector<std::string> unsorted);
bool isPositive(const std::string& s);
bool rightLength(const std::string& s);

//
// The main function.
int main(){
    int n;
    std::cin >> n;
    
    if (n >= 1 && n <= (2 * pow(10.0, 5.0))) createArray(n);
    
    return 0;
}

//
// This function allows the user to input data
// into the vector of positive integers.
void createArray(int n)
{
    std::vector<std::string> unsorted;
    
    std::string temp;
    for(int i = 0; i < n; i++) {
        std::cin >> temp;
        if (isPositive(temp) && rightLength(temp)) unsorted.push_back(temp);
        temp.clear();
    }
    
    std::sort(unsorted.begin(), unsorted.end(), compareBySizeAndValue);
    
    display(unsorted);
}

//
// This function iterates through the vector of positive integers.
void display(std::vector<std::string> unsorted)
{
    for(std::vector<std::string>::iterator it = unsorted.begin(); it != unsorted.end(); it++)
    {
        std::cout << *it << std::endl;
    }
}

//
// This boolean function is a predicate that is used
// by the std::sort function. It takes two std::strings
// as arguments. If the two strings are of the same length,
// then the function verifies which one is the smallest.
bool compareBySizeAndValue(const std::string &a, const std::string &b)
{
    return a.size() == b.size() ? a < b : a.size() < b.size();
}

//
// This boolean function makes sure that each string respects 
// this coding challenge's constraint about the length. The latter
// must be inclusively between 1 and 10 to the power of 6.
bool rightLength(const std::string& s) { return s.length() >= 1 && s.length() <= pow(10.0, 6.0); }

//
// This function makes sure that the std::string does not have
// one or many leading zeroes. Secondly, it makes sure that
// the std::string does not have a minus sign anywhere.
bool isPositive(const std::string& s) {
	return    !std::regex_search(s, std::regex("-")) &&
		      !std::regex_match(s, std::regex("0*"));
}

File: BigSortingSolution.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;

/**
 * The solution in Java 8 for the Big Sorting challenge
 * on HackerRank.com.
 * 
 * @author Anh Khoi Do
 * @version 2.0 2017-07-07
 */
public class Client {

    private static final String PATTERN = "^-0+[0-9]*";
    
    /**
     * The main method.
     * 
     * @param args The command line arguments.
     * @since 1.0 2017-07-05
     */
    public static void main(String[] args) {
        // Instance variable that determines the size of our string array.
        int n = 0;
        
        // Open an instance of the scanner class for the user's input.
        Scanner input = new Scanner(System.in);
        n = input.nextInt();
        
        // Constraint 1: the number of desired elements in the array (n)
        // must be inclusively between 1 and 2 * 10^5. If not, the program
        // will proceed towards the end of this method.
        if (n >= 1 && n <= (2 * Math.pow(10.0, 5.0))) {
            
            String[] temp = new String[n];
            String[] unsorted = new String[n];
            
            for (int i = 0; i < n; i++) {
                temp[i] = input.next();
                
                // Constraints 2 & 3: The temporary element's length must be inclusively 
                // between 1 and 10^6 before being inserted into the array named unsorted. 
                // Thirdly, the element must not have any leading zeroes or a minus sign.
                if ((temp[i].length() >= 1 && temp[i].length() <= Math.pow(10.0, 6.0)) &&
                	isPositiveNumber(temp[i])) unsorted[i] = temp[i];
                
            } // End of the for loop.
            
            // Sort the array of strings by using a Comparator<String>.
            // The latter will make sure that the strings are sorted the way
            // numbers will be. This will also prevent the program from sorting
            // the strings merely by characters.
            Arrays.sort(unsorted, new Comparator<String>() {
               @Override
               public int compare(String lhs, String rhs) {
                   int result = 0;
                   
                   //
                   // First if statement: If the two strings passed as arguments of the compare
                   // method are of the same length, then verify which one is the smallest.
                   //
                   // Else-if statement: If the left string argument has a bigger length than
                   // the right one, than move it further in the ascending order.
                   //
                   // Else: If the right string argument has a bigger length, than proceed
                   // to the other strings in the array.
                   if (lhs.length() == rhs.length()) result = lhs.compareTo(rhs);
                   else if (lhs.length() > rhs.length()) result = 1;
                   else result = -1;
                   
                   return result;
               }
            });
            
            for (int j = 0; j < n; j++) System.out.println(unsorted[j]);
            
        } // End of the if statement.
        
        // Close the instance of the Scanner class named input.
        input.close();
    }


    /**
     * This private static boolean method makes sure the string represents
     * a positive number and has no leading zeroes.
     *
     * @param s The {@link String} analyzed by the program.
     * @return true The method returns true if the {@link String} named s has no leading zeroes and represents a positive integer.
     * @since 2.0 2017-07-07
     */
    private static boolean isPositiveNumber(String s)
    {
    	return !s.contains("-") && (s.charAt(0) != '0') && !s.matches(PATTERN);
    }
}
